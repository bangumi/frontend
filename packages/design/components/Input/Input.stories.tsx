import type { Meta, StoryFn } from '@storybook/react';
import type { FC } from 'react';
import React from 'react';

import { Search as SearchIcon, UserLogin } from '@bangumi/icons';

import Select from '../Select';
import type { InputGroupProps, InputProps } from '.';
import Input from '.';

const componentMeta: Meta<typeof Input> = {
  title: 'modern/Input',
  component: Input,
  subcomponents: { 'Input.Group': Input.Group as FC<unknown> },
};

export default componentMeta;

const Template: StoryFn<InputProps & React.RefAttributes<HTMLInputElement>> = (args) => {
  return <Input {...args} />;
};

export const Default = Template.bind({});

Default.args = {
  placeholder: '你的 Email 地址',
};

const loginStyle = {
  width: 320,
};

export const IconPrefix = Template.bind({});

IconPrefix.args = {
  placeholder: '你的 Email 地址',
  type: 'email',
  prefix: <UserLogin />,
  wrapperStyle: loginStyle,
};

export const TextPrefix = Template.bind({});

TextPrefix.args = {
  placeholder: '填写对方的 username',
  prefix: '收件人:',
  rounded: true,
};

export const Password = Template.bind({});

Password.args = {
  placeholder: '您的密码',
  type: 'password',
  wrapperStyle: loginStyle,
};

export const NavSearch = Template.bind({});

NavSearch.args = {
  // 正式需要详细样式
  prefix: (
    <select
      name='select'
      style={{ outline: 'none', border: 'none', color: '#AAA6A6' }}
      defaultValue='value1'
    >
      <option value='value1'>全部条目</option>
    </select>
  ),
  suffix: <SearchIcon style={{ flexShrink: 0 }} />,
  wrapperStyle: {
    width: 256,
    height: 32,
    padding: '3px 12px 5px 13px',
  },
};

export const Rounded = Template.bind({});
Rounded.args = {
  placeholder: '取个标题…',
  rounded: true,
};

const InputGroupTemplate = (args: InputGroupProps) => {
  return (
    <Input.Group {...args}>
      <Select options={[{ label: '你好', value: '你好' }]} defaultValue='你好' />
      <Input />
      <Input />
      <Input rounded />
    </Input.Group>
  );
};

export const InputGroup = InputGroupTemplate.bind({});
