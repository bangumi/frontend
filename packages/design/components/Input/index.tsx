import React, { forwardRef } from 'react';
import classnames from 'classnames';

export interface InputProps {
  /* 同原生 input 标签 `type` */
  type?: string;
  /* 同原生 `placeholder` */
  placeholder?: string;
  /* 外层 wrapper 的样式 */
  wrapperStyle?: React.CSSProperties;
  /* 外层 wrapper 的自定义类名 */
  wrapperClass?: string;
  /* 前缀 */
  prefix?: React.ReactNode;
  /* 后缀 */
  suffix?: React.ReactNode;
}

/* eslint-disable react/prop-types */
// https://github.com/jsx-eslint/eslint-plugin-react/issues/3140
const Input = forwardRef<HTMLInputElement, InputProps>(
  ({ type = 'text', wrapperStyle, wrapperClass, prefix, suffix, placeholder, ...rest }, ref) => {
    return (
      <div className={classnames('bgm-input__wrapper', wrapperClass)} style={wrapperStyle}>
        {prefix && prefix}
        <input type={type} className='bgm-input' placeholder={placeholder} ref={ref} {...rest} />
        {suffix && suffix}
      </div>
    );
  },
);

export default Input;
