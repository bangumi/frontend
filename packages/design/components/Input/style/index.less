@import '../../../theme/base';

.bgm-input {
  &__wrapper {
    display: flex;
    position: relative;
    box-sizing: border-box;
    align-items: center;
    justify-content: space-between;
    padding: 9px 17px;
    font-size: 14px;
    border: 2px solid @gray-10;
    border-radius: 19px;
  }

  font-size: 1em;
  outline-style: none;
  border: none;
  color: @gray-100;
  padding: 0;
  width: 100%;

  &::placeholder {
    color: @gray-10;
  }
}
