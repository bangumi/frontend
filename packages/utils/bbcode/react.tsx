import React from 'react';

import { convert } from './convert';
import { Parser } from './parser';
import type { CodeNodeTypes, NodeTypes, VNode } from './types';

/* eslint-disable no-useless-escape */
const toCamelCase = (s: string) => s.replace(/(\-\w)/g, (k) => k[1]?.toUpperCase() ?? '');

const joinClassName = (cls: string | string[]) => {
  if (Array.isArray(cls)) {
    return cls.join(' ');
  }
  return cls;
};

const convertToReactProps = (node: VNode, key?: React.Key) => {
  const { props, style, className } = node;

  const transformedStyle: Record<string, string> = style
    ? Object.entries(style).reduce((pre, [key, val]) => {
        return {
          ...pre,
          [toCamelCase(key)]: val,
        };
      }, {})
    : {};

  const transformedClassName = className !== undefined ? joinClassName(className) : undefined;

  const reactProps = {
    ...props,
    style: transformedStyle,
    className: transformedClassName,
    key,
  };

  return reactProps;
};

export const renderNode = (node: NodeTypes, key?: React.Key): React.ReactElement | string => {
  if (typeof node === 'string') {
    const splittedStr: Array<string | React.ReactElement> = node.split(/\n/g);

    if (splittedStr.length < 2) {
      return splittedStr[0] ?? '';
    }
    const children = splittedStr.reduce<Array<string | React.ReactElement>>((pre, curr, idx) => {
      if (idx > 0) {
        return [...pre, React.createElement('br', null), curr];
      }
      return [...pre, curr];
    }, []);
    return React.createElement(React.Fragment, { key }, ...children);
  }

  const { type, children } = node;

  if (children) {
    return React.createElement(
      type,
      convertToReactProps(node, key),
      ...children.map((child, idx) => renderNode(child, idx)),
    );
  }
  return React.createElement(type, convertToReactProps(node, key));
};

export const render = (rawStr: string): Array<React.ReactElement | string> => {
  const nodes: CodeNodeTypes[] = new Parser(rawStr).parse();

  return nodes.map((node, idx) => renderNode(convert(node, {}), idx));
};
