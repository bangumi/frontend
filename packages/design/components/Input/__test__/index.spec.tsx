import { render } from '@testing-library/react';
import React from 'react';

import Input from '../index';

describe('Input', () => {
  it('placeholder', () => {
    const { getByPlaceholderText } = render(<Input placeholder='Hello' />);
    expect(getByPlaceholderText('Hello')).toBeInTheDocument();
  });

  it('ref should forward to input tag', () => {
    const ref = React.createRef<HTMLInputElement>();
    render(<Input ref={ref} />);
    expect(ref.current?.tagName).toBe('INPUT');
  });

  it('prefix should be rendered', () => {
    const { getByText } = render(<Input prefix='$' />);
    expect(getByText('$')).toBeInTheDocument();
  });

  it('suffix should be rendered', () => {
    const { getByText } = render(<Input suffix='$' />);
    expect(getByText('$')).toBeInTheDocument();
  });

  it('disabled className', () => {
    const { container } = render(<Input disabled />);
    expect(container).toMatchSnapshot();
  });

  it('alignment', () => {
    const { container } = render(<Input align='right' />);
    expect(container).toMatchSnapshot();
  });

  it('rounded variant', () => {
    const { container } = render(<Input rounded />);
    expect(container).toMatchSnapshot();
  });
});
