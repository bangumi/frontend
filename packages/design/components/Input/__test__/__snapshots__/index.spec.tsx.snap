// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Input > alignment 1`] = `
<div>
  <div
    class="bgm-input__wrapper"
  >
    <input
      class="bgm-input bgm-input--align-right"
      type="text"
    />
  </div>
</div>
`;

exports[`Input > disabled className 1`] = `
<div>
  <div
    class="bgm-input__wrapper bgm-input__wrapper--disabled"
  >
    <input
      class="bgm-input"
      disabled=""
      type="text"
    />
  </div>
</div>
`;

exports[`Input > rounded variant 1`] = `
<div>
  <div
    class="bgm-input__wrapper bgm-input__wrapper--rounded"
  >
    <input
      class="bgm-input"
      type="text"
    />
  </div>
</div>
`;
